src/ansi-c/c_nondet_symbol_factory.cpp

```diff
@@ -476,6 +484,85 @@ symbol_factoryt::get_deferred_size(irep_idt symbol_name) const
   return optional_lookup(deferred_array_sizes, symbol_name);
 }
 
+bool symbol_factoryt::is_string_array(const exprt &expr) const
+{
+  if(expr.id() != ID_symbol)
+  {
+    return false;
+  }
+  auto const &symbol_expr = to_symbol_expr(expr);
+  return object_factory_params.should_be_treated_as_string(
+    get_symbol_base_name(symbol_expr));
+}
+
+void symbol_factoryt::gen_nondet_string_init(
+  code_blockt &assignments,
+  const exprt &expr,
+  std::size_t depth,
+  const symbol_factoryt::recursion_sett &recursion_set)
+{
+  auto const &array_type = to_array_type(expr.type());
+  PRECONDITION(array_type.subtype() == char_type());
+  auto const array_size = numeric_cast_v<mp_integer>(array_type.size());
+  DATA_INVARIANT(array_size > 0, "Arrays should have positive size");
+  auto array_value_at = [&](mp_integer index) {
+    return index_exprt{expr, from_integer(index, size_type()), char_type()};
+  };
+  for(auto index = mp_integer(0); index < array_size - 1; ++index)
+  {
+    gen_nondet_printable_char_init(
+      assignments, array_value_at(index), depth, recursion_set);
+  }
+  // last char should be a 0 to terminate a C string
+  assignments.add(
+    code_assignt(array_value_at(array_size - 1), from_integer(0, char_type())));
+}
+
+void symbol_factoryt::gen_nondet_printable_char_init(
+  code_blockt &assignments,
+  const exprt &expr,
+  std::size_t depth,
+  const symbol_factoryt::recursion_sett &recursion_set)
+{
+  PRECONDITION(expr.type() == char_type());
+  // generate code like this:
```

Comment:
This comment is really helpful, thanks - Adding a bit of 'why?' commentary would also be super helpful too :-)